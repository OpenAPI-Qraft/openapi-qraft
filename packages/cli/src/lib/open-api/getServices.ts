import camelCase from 'camelcase';
import * as console from 'console';
import micromatch from 'micromatch';

import { getContentMediaType } from './getContent.js';
import { getOperationName } from './getOperationName.js';
import {
  getServiceNamesByOperationEndpoint,
  ServiceBaseNameByEndpointOption,
} from './getServiceNamesByOperationEndpoint.js';
import { getServiceNamesByOperationTags } from './getServiceNamesByOperationTags.js';
import type { OpenAPISchemaType } from './OpenAPISchemaType.js';

export type ServiceBaseName = ServiceBaseNameByEndpointOption | 'tags';

export type Service = {
  name: string;
  variableName: string;
  typeName: string;
  fileBaseName: string;
  operations: ServiceOperation[];
};

export type ServiceOperation = {
  method:
    | 'get'
    | 'put'
    | 'post'
    | 'patch'
    | 'delete'
    | 'options'
    | 'head'
    | 'trace';
  path: string;
  name: string;
  description: string | undefined;
  summary: string | undefined;
  deprecated: boolean | undefined;
  mediaType: string | undefined;
  errors: Record<string, string | undefined>;
  success: Record<string, string | undefined>;
  parameters: Record<string, any> | undefined;
};

export const getServices = (
  openApiJson: OpenAPISchemaType,
  {
    postfixServices = 'Service',
    serviceNameBase = 'endpoint[0]',
  }: { postfixServices?: string; serviceNameBase?: ServiceBaseName } = {},
  servicesGlob = ['**']
) => {
  const paths = openApiJson.paths;

  const services = new Map<string, Service>();

  const isPathMatch = createServicePathMatch(servicesGlob);

  for (const path in paths) {
    if (!paths.hasOwnProperty(path)) continue;

    for (const method in paths[path]) {
      if (!paths[path].hasOwnProperty(method)) continue;
      if (!isPathMatch(path)) continue;

      if (!supportedMethod(method)) {
        console.warn(
          `The path "${path}" HTTP method "${method}" is not supported`
        );
        continue;
      }

      const methodOperation = paths[path][method];

      const { success, errors } = Object.entries(
        methodOperation.responses
      ).reduce(
        (acc, [statusCode, response]) => {
          if (!response.content || typeof response.content !== 'object') {
            return acc;
          }

          const statusType =
            statusCode !== 'default' && // See "default" response https://swagger.io/docs/specification/describing-responses/#default
            Number(statusCode) < 400
              ? 'success'
              : 'errors';

          return {
            ...acc,
            [statusType]: {
              ...acc[statusType],
              [statusCode]: getContentMediaType(response.content),
            },
          };
        },
        {} as Record<'errors' | 'success', Record<string, string | undefined>>
      );

      const serviceFallbackBaseName = 'Default';

      const serviceNames =
        serviceNameBase === 'tags'
          ? getServiceNamesByOperationTags(
              paths[path][method]?.tags,
              serviceFallbackBaseName
            )
          : getServiceNamesByOperationEndpoint(
              path,
              serviceNameBase,
              serviceFallbackBaseName
            );

      for (const name of serviceNames) {
        if (!services.has(name)) {
          services.set(name, {
            name: camelCase(name),
            variableName: `${camelCase(name, {
              preserveConsecutiveUppercase: false,
            })}${postfixServices}`,
            typeName: `${name}${postfixServices}`,
            fileBaseName: `${name}${postfixServices}`,
            operations: [],
          });
        }
      }

      for (const name of serviceNames) {
        services.get(name)!.operations.push({
          method,
          path,
          errors,
          name: getOperationName(path, method, methodOperation.operationId),
          description: methodOperation.description,
          summary: methodOperation.summary,
          deprecated: methodOperation.deprecated,
          parameters:
            Array.isArray(methodOperation.parameters) &&
            !methodOperation.parameters.length
              ? undefined
              : methodOperation.parameters,
          mediaType: methodOperation.requestBody?.content
            ? getContentMediaType(methodOperation.requestBody.content)
            : undefined,
          success: success,
        });
      }
    }
  }

  return Array.from(services.values());
};

/**
 * Create a function to match service paths
 * @param servicesGlob
 */
export const createServicePathMatch = (servicesGlob: string[]) => {
  const servicePathGlobs = servicesGlob.reduce<
    Record<'match' | 'ignore', string[]>
  >(
    (acc, glob) => {
      glob.startsWith('!')
        ? acc.ignore.push(glob.slice(1))
        : acc.match.push(glob);
      return acc;
    },
    {
      match: [],
      ignore: [],
    }
  );

  return function isServicePatchMatch(path: string) {
    return micromatch.isMatch(path, servicePathGlobs.match, {
      ignore: servicePathGlobs.ignore,
    });
  };
};

export const supportedMethod = (
  method: unknown
): method is (typeof supportedHTTPMethods)[number] =>
  supportedHTTPMethods.includes(
    method as (typeof supportedHTTPMethods)[number]
  );

const supportedHTTPMethods = Object.values({
  get: 'get',
  put: 'put',
  post: 'post',
  patch: 'patch',
  delete: 'delete',
  options: 'options',
  head: 'head',
  trace: 'trace',
} satisfies { [key in ServiceOperation['method']]: key });
