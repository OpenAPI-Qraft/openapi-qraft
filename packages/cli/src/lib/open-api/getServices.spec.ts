import { describe, expect, it } from 'vitest';

import openAPI from './__fixtures__/openapi.json' assert { type: 'json' };
import { getServices } from './getServices.js';

describe('getServices', () => {
  it('matches snapshot with default options', () => {
    expect(getServices(openAPI)).toMatchSnapshot();
  });

  it('matches snapshot with custom `postfixServices`', () => {
    expect(getServices(openAPI, { postfixServices: 'Srv' })).toMatchSnapshot();
  });

  it('matches snapshot with custom `servicesGlob`', () => {
    expect(
      getServices(
        openAPI,
        { serviceNameBase: 'endpoint[0]', postfixServices: 'Service' },
        ['/files/**']
      )
    ).toMatchSnapshot();
  });

  it('matches snapshot with "serviceNameBase: endpoint"', () => {
    expect(
      getServices(openAPI, { serviceNameBase: 'endpoint[0]' })
    ).toMatchSnapshot();
  });

  it('matches snapshot with "serviceNameBase: tags"', () => {
    expect(getServices(openAPI, { serviceNameBase: 'tags' })).toMatchSnapshot();
  });
});
