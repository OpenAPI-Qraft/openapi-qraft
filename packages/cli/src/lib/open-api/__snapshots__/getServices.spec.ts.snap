// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getServices > matches snapshot with "serviceNameBase: endpoint" 1`] = `
[
  {
    "fileBaseName": "ApprovalPoliciesService",
    "name": "approvalPolicies",
    "operations": [
      {
        "deprecated": undefined,
        "description": "Retrieve a specific approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "get",
        "name": "getApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Order by",
            "in": "query",
            "name": "items_order",
            "required": false,
            "schema": {
              "default": "asc",
              "description": "Order by",
              "items": {
                "$ref": "#/components/schemas/OrderEnum",
              },
              "type": "array",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Get an approval policy by ID",
      },
      {
        "deprecated": undefined,
        "description": "Delete an existing approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "delete",
        "name": "deleteApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Limit of records to delete",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "description": "Limit",
              "type": "number",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Delete an approval policy",
      },
      {
        "deprecated": undefined,
        "description": "Update an existing approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": "application/json",
        "method": "patch",
        "name": "patchApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Limit of records to patch",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "description": "Limit",
              "type": "number",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Update an approval policy",
      },
    ],
    "typeName": "ApprovalPoliciesService",
    "variableName": "approvalPoliciesService",
  },
  {
    "fileBaseName": "FilesService",
    "name": "files",
    "operations": [
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "405": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "get",
        "name": "getFiles",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "query",
            "name": "id__in",
            "required": true,
            "schema": {
              "items": {
                "format": "uuid",
                "type": "string",
              },
              "maxItems": 100,
              "type": "array",
            },
          },
          {
            "description": "Page number",
            "in": "query",
            "name": "page",
            "required": false,
            "schema": {
              "description": "Page number",
              "type": "string",
            },
          },
        ],
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Get a files by ID",
      },
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "default": "application/json",
        },
        "mediaType": "multipart/form-data",
        "method": "post",
        "name": "postFiles",
        "parameters": undefined,
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Upload a files by ID",
      },
    ],
    "typeName": "FilesService",
    "variableName": "filesService",
  },
]
`;

exports[`getServices > matches snapshot with "serviceNameBase: tags" 1`] = `
[
  {
    "fileBaseName": "DefaultService",
    "name": "default",
    "operations": [
      {
        "deprecated": undefined,
        "description": "Retrieve a specific approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "get",
        "name": "getApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Order by",
            "in": "query",
            "name": "items_order",
            "required": false,
            "schema": {
              "default": "asc",
              "description": "Order by",
              "items": {
                "$ref": "#/components/schemas/OrderEnum",
              },
              "type": "array",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Get an approval policy by ID",
      },
    ],
    "typeName": "DefaultService",
    "variableName": "defaultService",
  },
  {
    "fileBaseName": "ApprovalPoliciesService",
    "name": "approvalPolicies",
    "operations": [
      {
        "deprecated": undefined,
        "description": "Delete an existing approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "delete",
        "name": "deleteApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Limit of records to delete",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "description": "Limit",
              "type": "number",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Delete an approval policy",
      },
      {
        "deprecated": undefined,
        "description": "Update an existing approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": "application/json",
        "method": "patch",
        "name": "patchApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Limit of records to patch",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "description": "Limit",
              "type": "number",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Update an approval policy",
      },
    ],
    "typeName": "ApprovalPoliciesService",
    "variableName": "approvalPoliciesService",
  },
  {
    "fileBaseName": "MainService",
    "name": "main",
    "operations": [
      {
        "deprecated": undefined,
        "description": "Delete an existing approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "delete",
        "name": "deleteApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Limit of records to delete",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "description": "Limit",
              "type": "number",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Delete an approval policy",
      },
      {
        "deprecated": undefined,
        "description": "Update an existing approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": "application/json",
        "method": "patch",
        "name": "patchApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Limit of records to patch",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "description": "Limit",
              "type": "number",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Update an approval policy",
      },
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "405": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "get",
        "name": "getFiles",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "query",
            "name": "id__in",
            "required": true,
            "schema": {
              "items": {
                "format": "uuid",
                "type": "string",
              },
              "maxItems": 100,
              "type": "array",
            },
          },
          {
            "description": "Page number",
            "in": "query",
            "name": "page",
            "required": false,
            "schema": {
              "description": "Page number",
              "type": "string",
            },
          },
        ],
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Get a files by ID",
      },
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "default": "application/json",
        },
        "mediaType": "multipart/form-data",
        "method": "post",
        "name": "postFiles",
        "parameters": undefined,
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Upload a files by ID",
      },
    ],
    "typeName": "MainService",
    "variableName": "mainService",
  },
  {
    "fileBaseName": "FilesService",
    "name": "files",
    "operations": [
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "405": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "get",
        "name": "getFiles",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "query",
            "name": "id__in",
            "required": true,
            "schema": {
              "items": {
                "format": "uuid",
                "type": "string",
              },
              "maxItems": 100,
              "type": "array",
            },
          },
          {
            "description": "Page number",
            "in": "query",
            "name": "page",
            "required": false,
            "schema": {
              "description": "Page number",
              "type": "string",
            },
          },
        ],
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Get a files by ID",
      },
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "default": "application/json",
        },
        "mediaType": "multipart/form-data",
        "method": "post",
        "name": "postFiles",
        "parameters": undefined,
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Upload a files by ID",
      },
    ],
    "typeName": "FilesService",
    "variableName": "filesService",
  },
]
`;

exports[`getServices > matches snapshot with custom \`postfixServices\` 1`] = `
[
  {
    "fileBaseName": "ApprovalPoliciesSrv",
    "name": "approvalPolicies",
    "operations": [
      {
        "deprecated": undefined,
        "description": "Retrieve a specific approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "get",
        "name": "getApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Order by",
            "in": "query",
            "name": "items_order",
            "required": false,
            "schema": {
              "default": "asc",
              "description": "Order by",
              "items": {
                "$ref": "#/components/schemas/OrderEnum",
              },
              "type": "array",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Get an approval policy by ID",
      },
      {
        "deprecated": undefined,
        "description": "Delete an existing approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "delete",
        "name": "deleteApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Limit of records to delete",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "description": "Limit",
              "type": "number",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Delete an approval policy",
      },
      {
        "deprecated": undefined,
        "description": "Update an existing approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": "application/json",
        "method": "patch",
        "name": "patchApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Limit of records to patch",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "description": "Limit",
              "type": "number",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Update an approval policy",
      },
    ],
    "typeName": "ApprovalPoliciesSrv",
    "variableName": "approvalPoliciesSrv",
  },
  {
    "fileBaseName": "FilesSrv",
    "name": "files",
    "operations": [
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "405": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "get",
        "name": "getFiles",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "query",
            "name": "id__in",
            "required": true,
            "schema": {
              "items": {
                "format": "uuid",
                "type": "string",
              },
              "maxItems": 100,
              "type": "array",
            },
          },
          {
            "description": "Page number",
            "in": "query",
            "name": "page",
            "required": false,
            "schema": {
              "description": "Page number",
              "type": "string",
            },
          },
        ],
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Get a files by ID",
      },
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "default": "application/json",
        },
        "mediaType": "multipart/form-data",
        "method": "post",
        "name": "postFiles",
        "parameters": undefined,
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Upload a files by ID",
      },
    ],
    "typeName": "FilesSrv",
    "variableName": "filesSrv",
  },
]
`;

exports[`getServices > matches snapshot with custom \`servicesGlob\` 1`] = `
[
  {
    "fileBaseName": "FilesService",
    "name": "files",
    "operations": [
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "405": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "get",
        "name": "getFiles",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "query",
            "name": "id__in",
            "required": true,
            "schema": {
              "items": {
                "format": "uuid",
                "type": "string",
              },
              "maxItems": 100,
              "type": "array",
            },
          },
          {
            "description": "Page number",
            "in": "query",
            "name": "page",
            "required": false,
            "schema": {
              "description": "Page number",
              "type": "string",
            },
          },
        ],
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Get a files by ID",
      },
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "default": "application/json",
        },
        "mediaType": "multipart/form-data",
        "method": "post",
        "name": "postFiles",
        "parameters": undefined,
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Upload a files by ID",
      },
    ],
    "typeName": "FilesService",
    "variableName": "filesService",
  },
]
`;

exports[`getServices > matches snapshot with default options 1`] = `
[
  {
    "fileBaseName": "ApprovalPoliciesService",
    "name": "approvalPolicies",
    "operations": [
      {
        "deprecated": undefined,
        "description": "Retrieve a specific approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "get",
        "name": "getApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Order by",
            "in": "query",
            "name": "items_order",
            "required": false,
            "schema": {
              "default": "asc",
              "description": "Order by",
              "items": {
                "$ref": "#/components/schemas/OrderEnum",
              },
              "type": "array",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Get an approval policy by ID",
      },
      {
        "deprecated": undefined,
        "description": "Delete an existing approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "delete",
        "name": "deleteApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Limit of records to delete",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "description": "Limit",
              "type": "number",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Delete an approval policy",
      },
      {
        "deprecated": undefined,
        "description": "Update an existing approval policy.",
        "errors": {
          "401": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": "application/json",
        "method": "patch",
        "name": "patchApprovalPoliciesId",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "path",
            "name": "approval_policy_id",
            "required": true,
            "schema": {
              "format": "uuid",
              "type": "string",
            },
          },
          {
            "description": "Limit of records to patch",
            "in": "query",
            "name": "limit",
            "required": false,
            "schema": {
              "description": "Limit",
              "type": "number",
            },
          },
        ],
        "path": "/approval_policies/{approval_policy_id}",
        "success": {
          "200": "application/json",
        },
        "summary": "Update an approval policy",
      },
    ],
    "typeName": "ApprovalPoliciesService",
    "variableName": "approvalPoliciesService",
  },
  {
    "fileBaseName": "FilesService",
    "name": "files",
    "operations": [
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "405": "application/json",
          "422": "application/json",
          "default": "application/json",
        },
        "mediaType": undefined,
        "method": "get",
        "name": "getFiles",
        "parameters": [
          {
            "example": "2023-06-04",
            "in": "header",
            "name": "x-monite-version",
            "required": true,
            "schema": {
              "format": "date",
              "type": "string",
            },
          },
          {
            "in": "query",
            "name": "id__in",
            "required": true,
            "schema": {
              "items": {
                "format": "uuid",
                "type": "string",
              },
              "maxItems": 100,
              "type": "array",
            },
          },
          {
            "description": "Page number",
            "in": "query",
            "name": "page",
            "required": false,
            "schema": {
              "description": "Page number",
              "type": "string",
            },
          },
        ],
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Get a files by ID",
      },
      {
        "deprecated": undefined,
        "description": undefined,
        "errors": {
          "default": "application/json",
        },
        "mediaType": "multipart/form-data",
        "method": "post",
        "name": "postFiles",
        "parameters": undefined,
        "path": "/files",
        "success": {
          "200": "application/json",
        },
        "summary": "Upload a files by ID",
      },
    ],
    "typeName": "FilesService",
    "variableName": "filesService",
  },
]
`;
